src/libpmemfile-posix/write.c

```diff
@@ -164,55 +191,140 @@ pmemfile_pwritev_internal(PMEMfilepool *pfp,
+	 */
+	for (int i = 0; i < iovcnt; ++i) {
+		size_t len = iov[i].iov_len;
+
+		if ((pmemfile_ssize_t)len < 0)
+			len = SSIZE_MAX;
+
+		if ((pmemfile_ssize_t)(ret + len) < 0)
+			len = SSIZE_MAX - ret;
+
+		if (offset + len < offset) /* overflow check */
+			len = SIZE_MAX - offset;
+
+		if (len > 0)
+			vinode_write(pfp, vinode, offset, last_block,
+					iov[i].iov_base, len);
+
+		ret += len;
+		offset += len;
+
+		if (len != iov[i].iov_len)
+			break;
+	}
+	ASSERT(ret > 0);
+
+	struct pmemfile_time starttm = tm;
+	if (get_current_time(&tm)) {
+		error = errno;
+		goto end;
+	}
 
-			ret += len;
-			offset += len;
+	int64_t tm_diff = (tm.sec - starttm.sec) * 1000000000 +
+			tm.nsec - starttm.nsec;
+	bool slots_changed = false;
 
-			if (len != iov[i].iov_len)
-				break;
+	/*
+	 * In theory we don't have to do update mtime here. We have a write lock
+	 * on vinode, so file changes can't be observed by another thread.
+	 * However if we performed long write which took more than, let's say,
+	 * 1ms we can update mtime. Time taken to do this will be negligible
+	 * comparing to time spent copying data.
+	 */
+	bool mtime_updated = false;
+	if (tm_diff >= 1000000) {
+		mtime_slot = inode_next_mtime_slot(inode);
+		inode->mtime[mtime_slot] = tm;
+		pmemfile_persist(pfp, &inode->mtime[mtime_slot]);
+		mtime_updated = true;
+
+		inode->slots.bits.mtime = mtime_slot;
+		slots_changed = true;
+	}
+
+	if (offset > inode_get_size(inode)) {
+		/*
+		 * If mtime was not updated, then do it here, because
+		 * mtime < ctime may confuse some applications into thinking
+		 * only non-content related metadata changed.
+		 * There's no penalty in updating mtime here because we are
+		 * going to update slot info anyway.
+		 */
+		if (!mtime_updated) {
+			mtime_slot = inode_next_mtime_slot(inode);
+			inode->mtime[mtime_slot] = tm;
+			pmemfile_flush(pfp, &inode->mtime[mtime_slot]);
+			mtime_updated = true;
 		}
 
+		bool size_slot = inode_next_size_slot(inode);
+		inode->size[size_slot] = offset;
+		pmemfile_flush(pfp, &inode->size[size_slot]);
+
+		bool ctime_slot = inode_next_ctime_slot(inode);
+		inode->ctime[ctime_slot] = tm;
+		pmemfile_flush(pfp, &inode->ctime[ctime_slot]);
+
 		/*
-		 * Update metadata only when any of the buffer lengths
-		 * was != 0.
+		 * We will update slot info now, so all slots must be on
+		 * the medium. Issue sfence to wait for that.
 		 */
-		if (ret > 0) {
-			struct pmemfile_time tm;
-			tx_get_current_time(&tm);
+		pmemfile_drain(pfp);
 
-			if (offset > inode_get_size(inode)) {
-				inode_tx_set_size(inode, offset);
-				inode_tx_set_ctime(inode, tm);
-			}
+		union pmemfile_inode_slots slots = inode->slots;
+		slots.bits.ctime = ctime_slot;
+		slots.bits.size = size_slot;
+		slots.bits.mtime = mtime_slot;
+		/*
+		 * All slot infos must be updated using one store. Force it by
+		 * using atomic store.
+		 */
+		__atomic_store_n(&inode->slots.value, slots.value,
+				__ATOMIC_RELEASE);
```

Comment:
Specifying release semantics here is a bit misleading.
Someone taking a quick glance can think this is used for synchronization with some acquire load somewhere else.
But the fail-safety is achieved by having an fences around, issued by pmemobj. The release semantics does absolutely nothing useful here.
```
pmemfile_drain(pfp); // sfence before

__atomic_store_n(&inode->slots.value, slots.value, __ATOMIC_RELEASE);

if (slots_changed)
    pmemfile_persist(pfp, &inode->slots);  //sfence after
```

The point of this atomic store is only in being non-divisible, or "atomic"  -- for the lovers of greek.
So I suggest `__ATOMIC_RELAXED` here.

Reply:
And this is also why I asked for commenting where the fences are, I wasn't sure if this does release something, or not.

Reply:
Indeed. Fixed.